plugins/Files/js/sagas/helpers.js

```diff
@@ -186,13 +186,62 @@ export const parseDownloads = (downloads) =>
 			})(),
 			siapath: download.siapath,
 			name: Path.basename(download.siapath),
+			bytestransferred: download.received,
+			totalbytes: download.filesize,
 			progress: Math.floor((download.received / download.filesize) * 100),
 			destination: download.destination,
 			type: 'download',
 			starttime: download.starttime,
 		}))
 		.sortBy((download) => -download.starttime)
 
+export const buildTransferTimes = (previousTransferTimes, transfers) => {
+	// Need a finite lookback window, lest we leak memory. This is also the window over which
+	// we calculate our average speed. This number can be adjusted.
+	const lookbackCount = 5
+	return transfers.reduce((map, transfer) => {
+		const previousTransferTime = previousTransferTimes.get(transfer.siapath)
+		if (previousTransferTime) {
+			return map.set(transfer.siapath, {
+				timestamps: previousTransferTime.timestamps.concat(Date.now()).slice(-lookbackCount),
+				bytes: previousTransferTime.bytes.concat(transfer.bytestransferred).slice(-lookbackCount),
+			})
+		}
+		return map.set(transfer.siapath, {
+			timestamps: [Date.now()],
+			bytes: [transfer.bytestransferred],
+		})
+	}, Map())
+}
+
+const calculateSpeed = (transferTime) => {
```

Comment:
This should have a docstring, along with the previous func (`buildTransferTimes`), and the next (`addTransferSpeeds`). We're probably not completely consistent about this currently, but going forward I would like to see all UI helpers to have useful docstrings.